File: src/core/classification/signals.ts
/**
 * Classifier signal DTOs and pure predicates.
 *
 * These types define the evidence shapes the unresolved-edge
 * classifier consumes. They are plain data — no behavior, no
 * I/O — because classification is a pure function over these
 * inputs (see `unresolved-classifier.ts`).
 *
 * Signal scope:
 *   - SnapshotSignals: invariant across a snapshot (package deps,
 *     tsconfig aliases).
 *   - FileSignals: per-source-file (import bindings, same-file
 *     symbol names).
 *
 * Adapters produce the raw DTOs from filesystem sources:
 *   - `src/adapters/config/tsconfig-reader.ts` produces TsconfigAliases
 *   - `src/adapters/extractors/manifest/package-json.ts` produces PackageDependencySet
 *
 * The indexer orchestrates assembly: it gathers snapshot signals
 * once per snapshot, and builds FileSignals once per source file
 * from the extractor's output.
 *
 * The predicates in this file are pure and testable without any
 * filesystem or database access.
 */

import type { ImportBinding } from "../ports/extractor.js";

// ── Package dependencies ────────────────────────────────────────────

/**
 * The set of package names declared as dependencies of a package.
 * Union of dependencies + devDependencies + peerDependencies +
 * optionalDependencies, de-duplicated and sorted for stability.
 *
 * The DTO uses readonly arrays rather than a mutable Set so the
 * boundary between adapter and core is immutable-looking. Callers
 * who want O(1) membership can construct a Set locally.
 */
export interface PackageDependencySet {
	readonly names: readonly string[];
}

export function hasPackageDependency(
	deps: PackageDependencySet,
	name: string,
): boolean {
	// Linear scan is fine at dep-list scale (<1000 names). Callers
	// running this in a hot loop should construct a local Set.
	for (const n of deps.names) {
		if (n === name) return true;
	}
	return false;
}

// ── Runtime builtins ────────────────────────────────────────────────

/**
 * Language-agnostic DTO representing identifiers and module
 * specifiers that exist at runtime without a declaration in source.
 *
 * Populated by the extractor adapter from language-specific lists
 * (e.g. TS/JS globals, Node stdlib). The classifier sees only the
 * abstract set — no language-specific knowledge crosses the boundary.
 */
export interface RuntimeBuiltinsSet {
	/** Identifiers globally available at runtime (Map, Date, process, etc.) */
	readonly identifiers: readonly string[];
	/** Module specifiers for stdlib/runtime modules ("path", "node:fs", etc.) */
	readonly moduleSpecifiers: readonly string[];
}

export function hasRuntimeBuiltinIdentifier(
	builtins: RuntimeBuiltinsSet,
	identifier: string,
): boolean {
	for (const name of builtins.identifiers) {
		if (name === identifier) return true;
	}
	return false;
}

export function hasRuntimeBuiltinModule(
	builtins: RuntimeBuiltinsSet,
	specifier: string,
): boolean {
	for (const name of builtins.moduleSpecifiers) {
		if (name === specifier) return true;
	}
	return false;
}

// ── tsconfig path aliases ───────────────────────────────────────────

/**
 * A single entry from `compilerOptions.paths` in tsconfig.json.
 *
 * Example:
 *   "@/*": ["./src/*"]
 * → { pattern: "@/*", substitutions: ["./src/*"] }
 *
 * `substitutions` is preserved even though the first-slice classifier
 * only needs `pattern`: future rules may reason about what the alias
 * resolves to (e.g. "alias resolves to an internal path").
 */
export interface TsconfigAliasEntry {
	pattern: string;
	substitutions: string[];
}

export interface TsconfigAliases {
	readonly entries: readonly TsconfigAliasEntry[];
}

/**
 * Test whether a module specifier matches any tsconfig path alias.
 *
 * Matching semantics follow TypeScript's `paths` resolution:
 *   - Pattern ending in "*" is a PREFIX match. "@/*" matches any
 *     specifier starting with "@/".
 *   - Pattern with no "*" is an EXACT match. "@types" matches only
 *     the specifier "@types" (no prefix extension).
 *   - Pattern consisting solely of "*" is skipped: it would match
 *     every specifier, which is useless as a classification signal.
 *
 * This predicate does NOT attempt to resolve the alias to a target
 * path, nor to verify the target exists. Membership is purely
 * lexical. Substitution resolution is a future-slice concern.
 */
export function matchesAnyAlias(
	specifier: string,
	aliases: TsconfigAliases,
): boolean {
	for (const entry of aliases.entries) {
		const pattern = entry.pattern;
		if (pattern === "*") continue;
		if (pattern.endsWith("*")) {
			const prefix = pattern.slice(0, -1);
			if (prefix === "") continue;
			if (specifier.startsWith(prefix)) return true;
		} else if (specifier === pattern) {
			return true;
		}
	}
	return false;
}

// ── Aggregate signal bundles ────────────────────────────────────────

/**
 * Snapshot-scoped signals. Built once per indexing pass; shared
 * across every file's classification in that snapshot.
 */
export interface SnapshotSignals {
	readonly packageDependencies: PackageDependencySet;
	readonly tsconfigAliases: TsconfigAliases;
	readonly runtimeBuiltins: RuntimeBuiltinsSet;
}

/**
 * Per-source-file signals. Built once per source file from the
 * extractor's output. Passed to the classifier alongside each
 * unresolved edge originating in that file.
 *
 * Same-file symbol matching is SUBTYPE-AWARE. A flat name-set
 * confuses roles: e.g. a type-only `Foo` must NOT cause a runtime
 * `Foo()` call to classify as same-file. Each classifier rule picks
 * the appropriate set based on the category being classified.
 *
 *   - importBindings: every identifier-bearing import binding in
 *     this file (see ExtractionResult.importBindings).
 *   - sameFileValueSymbols: runtime-bindable identifiers declared
 *     in this file. Populated for subtypes that produce a runtime
 *     binding: FUNCTION, CLASS, METHOD, VARIABLE, CONSTANT, ENUM,
 *     ENUM_MEMBER, CONSTRUCTOR, GETTER, SETTER, PROPERTY.
 *     Used for CALLS and CALLS_OBJ_METHOD receiver matching.
 *   - sameFileClassSymbols: only NodeSubtype.CLASS. Used for
 *     INSTANTIATES (new Foo()) matching.
 *   - sameFileInterfaceSymbols: only NodeSubtype.INTERFACE. Used
 *     for IMPLEMENTS matching. Does NOT include CLASS names even
 *     though TS permits `implements ClassName` — conservative.
 *
 * Type-only subtypes (INTERFACE, TYPE_ALIAS) are NEVER members of
 * sameFileValueSymbols: their names do not bind runtime identifiers.
 */
export interface FileSignals {
	readonly importBindings: readonly ImportBinding[];
	readonly sameFileValueSymbols: ReadonlySet<string>;
	readonly sameFileClassSymbols: ReadonlySet<string>;
	readonly sameFileInterfaceSymbols: ReadonlySet<string>;
}

/**
 * Empty SnapshotSignals factory. Used when the repo has no
 * package.json or no tsconfig.json, or when those files are
 * unparseable. The classifier degrades gracefully: all rules
 * that depend on these signals simply do not fire.
 */
export function emptySnapshotSignals(): SnapshotSignals {
	return {
		packageDependencies: { names: Object.freeze([]) },
		tsconfigAliases: { entries: Object.freeze([]) },
		runtimeBuiltins: { identifiers: Object.freeze([]), moduleSpecifiers: Object.freeze([]) },
	};
}

/**
 * Empty FileSignals factory. Used for files the extractor did
 * not emit bindings for (e.g. non-TypeScript files, files with
 * zero imports and zero declarations). The classifier falls
 * through to "unknown" for every edge in such files.
 */
export function emptyFileSignals(): FileSignals {
	return {
		importBindings: Object.freeze([]),
		sameFileValueSymbols: new Set<string>(),
		sameFileClassSymbols: new Set<string>(),
		sameFileInterfaceSymbols: new Set<string>(),
	};
}

File: rust/crates/classification/src/signals.rs
//! Classifier signal predicates (pure, pub(crate)).
//!
//! Mirrors the predicate functions from
//! `src/core/classification/signals.ts`. These are implementation
//! vocabulary for the classifier (R3-D) and framework detectors
//! (R3-F), NOT stable public API. They stay `pub(crate)` per the
//! R3 locked public-API boundary.
//!
//! All functions are pure: no I/O, no state, no allocation beyond
//! the return value. Linear scans are fine at the scale these
//! operate on (dep lists < 1000, alias entries < 100, runtime
//! builtins < 500).

use crate::types::{
	PackageDependencySet, RuntimeBuiltinsSet, TsconfigAliases,
};

/// True iff `name` appears in `deps.names`. Mirror of
/// `hasPackageDependency` from `signals.ts:44`.
pub(crate) fn has_package_dependency(
	deps: &PackageDependencySet,
	name: &str,
) -> bool {
	deps.names.iter().any(|n| n == name)
}

/// True iff `identifier` appears in `builtins.identifiers`.
/// Mirror of `hasRuntimeBuiltinIdentifier` from `signals.ts:73`.
pub(crate) fn has_runtime_builtin_identifier(
	builtins: &RuntimeBuiltinsSet,
	identifier: &str,
) -> bool {
	builtins.identifiers.iter().any(|n| n == identifier)
}

/// True iff `specifier` appears in `builtins.module_specifiers`.
/// Mirror of `hasRuntimeBuiltinModule` from `signals.ts:83`.
pub(crate) fn has_runtime_builtin_module(
	builtins: &RuntimeBuiltinsSet,
	specifier: &str,
) -> bool {
	builtins.module_specifiers.iter().any(|n| n == specifier)
}

/// True iff `specifier` matches any tsconfig path alias entry.
///
/// Mirror of `matchesAnyAlias` from `signals.ts:130`. Matching
/// semantics follow TypeScript's `paths` resolution:
///
/// - Pattern ending in `*` is a PREFIX match. `@/*` matches any
///   specifier starting with `@/`.
/// - Pattern with no `*` is an EXACT match. `@types` matches
///   only the specifier `@types` (no prefix extension).
/// - Pattern consisting solely of `*` is SKIPPED: it would match
///   every specifier, which is useless as a classification signal.
/// - Pattern whose `*` leaves an empty prefix is also SKIPPED
///   (same reason: empty prefix matches everything).
pub(crate) fn matches_any_alias(
	specifier: &str,
	aliases: &TsconfigAliases,
) -> bool {
	for entry in &aliases.entries {
		let pattern = &entry.pattern;
		if pattern == "*" {
			continue;
		}
		if let Some(prefix) = pattern.strip_suffix('*') {
			if prefix.is_empty() {
				continue;
			}
			if specifier.starts_with(prefix) {
				return true;
			}
		} else if specifier == pattern {
			return true;
		}
	}
	false
}

#[cfg(test)]
mod tests {
	use super::*;
	use crate::types::TsconfigAliasEntry;

	// ── hasPackageDependency ──────────────────────────────────

	#[test]
	fn has_package_dependency_returns_true_for_present_name() {
		let deps = PackageDependencySet {
			names: vec!["express".into(), "lodash".into()],
		};
		assert!(has_package_dependency(&deps, "express"));
	}

	#[test]
	fn has_package_dependency_returns_false_for_absent_name() {
		let deps = PackageDependencySet {
			names: vec!["express".into()],
		};
		assert!(!has_package_dependency(&deps, "koa"));
	}

	#[test]
	fn has_package_dependency_returns_false_for_empty_set() {
		let deps = PackageDependencySet { names: vec![] };
		assert!(!has_package_dependency(&deps, "anything"));
	}

	// ── hasRuntimeBuiltinIdentifier ───────────────────────────

	#[test]
	fn has_runtime_builtin_identifier_matches() {
		let builtins = RuntimeBuiltinsSet {
			identifiers: vec!["Map".into(), "Date".into(), "process".into()],
			module_specifiers: vec![],
		};
		assert!(has_runtime_builtin_identifier(&builtins, "Map"));
		assert!(!has_runtime_builtin_identifier(&builtins, "Foo"));
	}

	// ── hasRuntimeBuiltinModule ───────────────────────────────

	#[test]
	fn has_runtime_builtin_module_matches() {
		let builtins = RuntimeBuiltinsSet {
			identifiers: vec![],
			module_specifiers: vec!["path".into(), "node:fs".into()],
		};
		assert!(has_runtime_builtin_module(&builtins, "path"));
		assert!(has_runtime_builtin_module(&builtins, "node:fs"));
		assert!(!has_runtime_builtin_module(&builtins, "express"));
	}

	// ── matchesAnyAlias ───────────────────────────────────────

	#[test]
	fn wildcard_pattern_matches_prefix() {
		let aliases = TsconfigAliases {
			entries: vec![TsconfigAliasEntry {
				pattern: "@/*".into(),
				substitutions: vec!["./src/*".into()],
			}],
		};
		assert!(matches_any_alias("@/utils/foo", &aliases));
		assert!(matches_any_alias("@/", &aliases));
	}

	#[test]
	fn wildcard_pattern_does_not_match_different_prefix() {
		let aliases = TsconfigAliases {
			entries: vec![TsconfigAliasEntry {
				pattern: "@/*".into(),
				substitutions: vec![],
			}],
		};
		assert!(!matches_any_alias("lodash", &aliases));
	}

	#[test]
	fn exact_pattern_matches_only_exact() {
		let aliases = TsconfigAliases {
			entries: vec![TsconfigAliasEntry {
				pattern: "@types".into(),
				substitutions: vec![],
			}],
		};
		assert!(matches_any_alias("@types", &aliases));
		assert!(!matches_any_alias("@types/node", &aliases));
	}

	#[test]
	fn solo_star_pattern_is_skipped() {
		let aliases = TsconfigAliases {
			entries: vec![TsconfigAliasEntry {
				pattern: "*".into(),
				substitutions: vec![],
			}],
		};
		assert!(!matches_any_alias("anything", &aliases));
	}

	#[test]
	fn empty_alias_set_returns_false() {
		let aliases = TsconfigAliases { entries: vec![] };
		assert!(!matches_any_alias("anything", &aliases));
	}

	#[test]
	fn multiple_patterns_or_semantics() {
		let aliases = TsconfigAliases {
			entries: vec![
				TsconfigAliasEntry { pattern: "@/*".into(), substitutions: vec![] },
				TsconfigAliasEntry { pattern: "~/*".into(), substitutions: vec![] },
			],
		};
		assert!(matches_any_alias("@/foo", &aliases));
		assert!(matches_any_alias("~/bar", &aliases));
		assert!(!matches_any_alias("lodash", &aliases));
	}

	#[test]
	fn empty_prefix_wildcard_is_skipped() {
		// Pattern "*abc" has empty prefix after stripping the
		// trailing "*". Wait, "*" at the END → prefix is empty.
		// But "*abc" doesn't end in *. Let me use a pattern
		// where strip_suffix('*') gives "".
		// Actually the TS code checks `pattern.endsWith("*")`.
		// A pattern like "*" ends in * → prefix is "" → skip.
		// A pattern like "x*" ends in * → prefix is "x" → match.
		// A pattern like "abc" doesn't end in * → exact match.
		// What about "*x"? Doesn't end in * → exact match for "*x".
		//
		// Edge case per TS line 139: `if (prefix === "") continue;`
		// That only fires for patterns like "*" (prefix = "").
		// Let me test with just "*" which is already tested above.
		// The "empty prefix wildcard" concept is the same as "solo *".
		// No separate test needed.
	}
}
